worker/src/services/IngestionService/index.ts

```diff
@@ -0,0 +1,788 @@
+  TraceEventType,
+  traceRecordInsertSchema,
+  TraceRecordInsertType,
+  traceRecordReadSchema,
+} from "@langfuse/shared/src/server";
+
+import { tokenCount } from "../../features/tokenisation/usage";
+import { instrumentAsync } from "../../instrumentation";
+import logger from "../../logger";
+import { IngestionFlushQueue } from "../../queues/ingestionFlushQueue";
+import { convertJsonSchemaToRecord, overwriteObject } from "./utils";
+
+enum TableName {
+  Traces = "traces",
+  Scores = "scores",
+  Observations = "observations",
+}
+
+enum EntityType {
+  Trace = "trace",
+  Score = "score",
+  Observation = "observation",
+  SDK_LOG = "sdk-log",
+}
+
+const immutableEntityKeys: {
+  [TableName.Traces]: (keyof TraceRecordInsertType)[];
+  [TableName.Scores]: (keyof ScoreRecordInsertType)[];
+  [TableName.Observations]: (keyof ObservationRecordInsertType)[];
+} = {
+  [TableName.Traces]: ["id", "project_id", "timestamp", "created_at"],
+  [TableName.Scores]: [
+    "id",
+    "project_id",
+    "timestamp",
+    "trace_id",
+    "created_at",
+  ],
+  [TableName.Observations]: [
+    "id",
+    "project_id",
+    "trace_id",
+    "start_time",
+    "created_at",
+  ],
+};
+
+export class IngestionService {
+  constructor(
+    private redis: Redis,
+    private prisma: PrismaClient,
+    private ingestionFlushQueue: IngestionFlushQueue,
+    private bufferTtlSeconds: number
+  ) {}
+
+  public async addBatch(
+    events: IngestionBatchEventType,
+    projectId: string
+  ): Promise<void> {
+    const ingestedEvents = events.map(async (event) => {
+      if (!("id" in event.body && event.body.id)) {
+        logger.warn(
+          `Received ingestion event without id, ${JSON.stringify(event)}`
+        );
+
+        return null;
+      }
+
+      const projectEntityKey = this.getProjectEntityKey({
+        entityId: event.body.id,
+        eventType: this.getEventType(event),
+        projectId,
+      });
+      const serializedEventData = JSON.stringify({ ...event, projectId });
+
+      await this.redis.lpush(projectEntityKey, serializedEventData);
+      await this.redis.expire(projectEntityKey, this.bufferTtlSeconds);
+      await this.ingestionFlushQueue.add(QueueJobs.FlushIngestionEntity, null, {
+        jobId: projectEntityKey,
+      });
+    });
+
+    await Promise.all(ingestedEvents);
+  }
+
+  public async flush(projectEntityKey: string): Promise<void> {
+    const eventList = (await this.redis.lrange(projectEntityKey, 0, -1))
+      .map((serializedEventData) => {
+        const parsed = ingestionEventWithProjectId.safeParse(
+          JSON.parse(serializedEventData)
+        );
+
+        if (!parsed.success) {
+          logger.error(
+            `Failed to parse event ${serializedEventData} : ${parsed.error}`
+          );
+
+          return null;
+        }
+
+        return parsed.data;
+      })
+      .filter(Boolean) as IngestionEventWithProjectIdType[];
+
+    if (eventList.length === 0) {
+      throw new Error(
```

Comment:
We should not throw here. If we have one creation and one update event, the second one could read an empty queue. We should log this and handle gracefully.

Reply:
Yes, you are right.